### ISSUE TITLE
false positive on arrow function overloads that return a generic with a conditional type that has `never` in it

### ISSUE THREAD

**Issue by DetachHead** [microsoft/TypeScript#45264]
# Bug Report

### 🔎 Search Terms

arrow function overload

### 🕗 Version & Regression Information
4.4.0-dev.20210801

### ⏯ Playground Link

[Playground link with relevant code](https://www.typescriptlang.org/play?jsx=0&ts=4.4.0-dev.20210801#code/C4TwDgpgBAQghgJwDwBUB8UC8UVQgD2AgDsATAZygAYoB+KYiANwgSgC4GBXAWwCNWAWABQIgPRiU4aAHJivAQhlQAlpWIB7YFDjlyKgObE4fADbRgGqKEhQ5zVjIB0ACgBMAZjduAlCIDGGsTk2gBmGhqcAN4iUHFQLj6c8MhUaLHxqGguTHCmXBCcKEmwiFkZcVk5eQW0RSUp5cIAvlhQVbn5EHU4DWXoWBgxwvFQCBDAXAjEUJ0FOuoOCCLNIuJikAihEP7ApiBQoSqMUEEMGgg8eYdcxLsqQeQiobf3Z3yIicllac+vwA8Zh9kOhql16t8QelhC87gD3v1snNuhDSlCoMNRuNJtNZjVoLoGEsVkA)

### 💻 Code

```ts
type Bar<T> = T extends 0 ? never : number

//Type 'number' is not assignable to type 'never'.(2322)
const foo: {
    (): Bar<0>
    <T>(value: T): Bar<T>
    <T>(value?: T): Bar<T>
} = <T>(value?: T): Bar<T> => {
    return value as never
}

//perfectly fine on normal functions
function bar(): Bar<0>
function bar<T>(value: T): Bar<T>
function bar<T>(value?: T): Bar<T> {
    return value as never
}
```

### 🙁 Actual behavior

```
Type '<T>(value?: T | undefined) => Bar<T>' is not assignable to type '{ (): never; <T>(value: T): Bar<T>; <T>(value?: T | undefined): Bar<T>; }'.
  Type 'number' is not assignable to type 'never'.(2322)
```

### 🙂 Expected behavior

no error

**Comment by jcalz** [comment]
This doesn't have much or anything to do with generics, conditional types, or `never`.  Overloaded function statements are checked more loosely than the equivalent overloaded arrow function:

```ts
// error, 1 is not assignable to 0
const f: { (x: 0): 0; (x: 1): 1 } =
    (x: 0 | 1) => { return x }

// no error
function g(x: 0): 0;
function g(x: 1): 1;
function g(x: 0 | 1) {
    return x;
}
```
[Playground link to code](https://www.typescriptlang.org/play?jsx=0&ts=4.4.0-dev.20210801#code/MYewdgzgLgBAZgLhgbxgCgB5IAwEocDc6WMAjPmTAL4wC8AsAFAwvE4wA+ZudAfCjABOAUygBXQWBgZqTJnDFhgUAJbgYAc0w4K2AvMXK1UrSXJJS+xgqWr1p9l3IomrIaIlSMVqkA)

So this looks like a duplicate of #33482.

**Comment by DetachHead** [comment]
so if my understanding is correct, it's invalid because there's no signature for the scenario where the input could be `0` and it returns `1`? (which i don't want to allow for anyway)

in that case, the following should be valid but it still fails:
```ts
const f: { (x: 0): 0; (x: 1): 1; (x: 1 | 0): 1 | 0 } =
    (x: 0 | 1) => { return x }
```

**Comment by jcalz** [comment]
No, it's invalid because the compiler is unable verify that the implementation is compatible with *each* call signature.  It sees the implementation as being of type `(x: 0 | 1) => 0 | 1`, meaning that for all it knows, you could pass in a `0` and get a `1` out.  (See [this comment](https://github.com/microsoft/TypeScript/issues/35338#issuecomment-558867975) on another duplicate of #33482.)  The type `(x: 0 | 1) => 0 | 1` is therefore not assignable to `(x: 0) => 0` and so the assignment fails immediately, no matter what happens with other call signatures.  

The compiler never really performs the analysis necessary to *accurately* type check overloaded functions.  Arrow functions tend to give false positives because the compiler is too cautious (which is #33482), while overloaded function statements tend to give false negatives because the compiler is too lax (which is #13235).

**Comment by DetachHead** [comment]
ah i get it now, thanks.

so the solution is to do it like this
```ts
const f: { (x: 0): 0; (x: 1): 1; } =
  <T extends 0 | 1>(x: T): T => { return x }
  ```
